src/librustc_metadata/locator.rs

```diff
@@ -681,12 +683,21 @@ impl<'a> CrateLocator<'a> {
         let mut ret: Option<(PathBuf, PathKind)> = None;
         let mut error = 0;
 
+        // If we are producing an rlib, and we've already loaded metadata, then
+        // we should not attempt to discover further crate sources. This means
+        // that under -Zbinary-dep-depinfo we will not emit a dependency edge on
+        // the *unused* rlib, and by returning `None` here immediately we
+        // guarantee that we do indeed not use it.
+        //
+        // See also #68149 which provides more detail on why emitting the
+        // dependency on the rlib is a bad thing.
+        //
+        // We currenty do not verify that these other sources are even in sync,
+        // and this is arguably a bug (see #10786), but because reading metadata
+        // is quite slow (especially from dylibs) we currently do not read it
+        // from the other crate sources.
         if slot.is_some() {
-            // FIXME(#10786): for an optimization, we only read one of the
-            //                libraries' metadata sections. In theory we should
-            //                read both, but reading dylib metadata is quite
-            //                slow.
-            if m.is_empty() {
+            if m.is_empty() || self.sess.crate_types.borrow() == &[config::CrateType::Rlib] {
```

Comment:
Hmm, we don't have a separate `CrateType::Rmeta`.
If the build is supposed to produce a `rmeta` file and stop we should act in the same way and return `None` regardless of `crate_types`.
I.e. this logic needs to also consider `sess.opts.output_types` somehow (which can be factored out into `Session::we_only_need_metadata_not_object_code`).